### ISSUE TITLE
Support inter-Pod affinity to one or more Pods

### ISSUE THREAD

**Issue by bsalamat** [kubernetes/kubernetes#68701]
In the current implementation of inter-Pod affinity, the scheduler looks for a single existing pod that can satisfy all the terms of inter-pod affinity of an incoming pod.
With the recent changes made to the implementation of inter-Pod affinity, we can now support multiple pods satisfying inter-pod affinity. One of the main reasons we didn't pursue the idea before was the fact that the inter-pod affinity feature was very slow (3 orders of magnitude slower than other scheduler predicates). We didn't want to add more complication to an already slow predicate. However, we can now think about adding the feature.
With this feature, a pod can have multiple affinity terms satisfied by a group of pods, as opposed to only a single pod. For example:

```
- Assume that the cluster has two nodes: 
- nodeA located in in zone1/region1
- nodeB located in zone2/region1

- There are two existing pods on these nodes:
- Pod1:
    - nodeName: "nodeA"
    - label "foo":""

- Pod2:
    - nodeName: "nodeB"
    - label "bar":""

- Pod3 comes in with inter-pod affinity:
    - affinity terms:
        - {label "foo" exists, topologyKey: "region"}
        - {label "bar" exists, topologyKey: "zone"}
```

With our current (K8s 1.12) implementation, Pod3 is not schedulable, because there is no single pod that satisfies all of its affinity terms. However, if we support multiple pods satisfying the affinity terms, Pod3 can be scheduled on nodeB. Pod1 satisfies the first term of its affinity in region1 and Pod2 satisfies its second term in zone2. So, any node in zone2/region1 will be feasible for Pod3.

Given the current implementation of inter-pod affinity and the use of "Topology Pair Maps", I believe implementing this feature requires little changes and won't have noticeable performance impact.

/kind feature
/sig scheduling

cc/ @Huang-Wei @ahmad-diaa

**Comment by bsalamat** [comment]
It is worth noting that this will apply only to inter-pod affinity, not inter-pod anti-affinity. Inter-pod anti-affinity is considered "violated" if there a pod that matches ANY terms of the anti-affinity. So, matching against a group of Pods does not make sense for anti-affinity.

**Comment by Huang-Wei** [comment]
@bsalamat I have a full picture on affinity/anti-affinity code after delivering #68173. I'm more than happy to help on this :)

/assign

**Comment by Huang-Wei** [comment]
Re @misterikkit:

> a canary workload is launched with pod-affinity for labels {app="foo", env="canary"}. That workload could end up in a topology containing {app="foo", env="prod"} & {app="bar", env="canary"} after this change.

If the goal is to gather pods with labels {app="foo", env="canary"}, the workload should have it defined within the same affinityTerm, in different expressions.

And if app="foo" and env="canary" are defined in two different affinityTerms, then yes, after the change, a topology containing {app="foo", env="prod"} & {app="bar", env="canary"} can be a fit.

**Comment by alculquicondor** [comment]
/reopen

We might be ready for this, performance-wise.

However, this is change of semantics that I'm not sure plays well with API deprecation policies.

**Comment by Huang-Wei** [comment]
> However, this is change of semantics that I'm not sure plays well with API deprecation policies.

The API doesn't give a clear explanation on it in my eyes.

And given that PodAffinity has been there for 10+ releases, to be honest I guess rare users have noticed this implicit semantics... or at lease not care too much.

**Comment by alculquicondor** [comment]
Maybe we can add this in 1.20

**Comment by sanposhiho** [comment]
/reopen
/remove-lifecycle rotten

Seems still valid. Even if we don't implement it, the current behavior should be documented at least.

**Comment by sanposhiho** [comment]
@Huang-Wei I can take over it if you like.

**Comment by sanposhiho** [comment]
I'm not sure if changing the current inter-Pod affinity behavior is allowed or not. But if we should avoid changing the current multiple inter-Pod affinity behavior so that we won't break existing clusters, I guess we can add a new field like `multiplePodAffinity`?

**Comment by Huang-Wei** [comment]
The diff would be, upon multiple podAffinity constraints, a Pod failed scheduling before could be scheduled. So I read this diff as a net benefit - nobody would spend time crafting podAffinity and wish it cannot be scheduled, right? With that regard, I think it should be allowed.

**Comment by Huang-Wei** [comment]
It's just this requirement isn't requested by any end-user over the years. So even if we decide to implement it, given there may involve non-trivial changes, it would be pretty low-priority.

**Comment by sanposhiho** [comment]
We've got one request for this [here](https://kubernetes.slack.com/archives/C09TP78DV/p1652496371446289)

**Comment by alculquicondor** [comment]
This change is not trivial from an API perspective, so it needs a KEP. And we are too close to the enhancements freeze. I suggest to leave it for 1.27.

**Comment by sanposhiho** [comment]
Opened https://github.com/kubernetes/enhancements/issues/3604.
I think we have two options for it.

### 1. a new enum field to decide how the scheduler evaluate multiple PodAffinity 

Introduce a new enum field in [`PodAffinity`](https://pkg.go.dev/k8s.io/api/core/v1#PodAffinity) and let users choose the way how to evaluate PodAffnities.

The field name would be like `satisfacatableBy` (value: `multiplePods` and `singlePod`(default))...?  (I cannot come up with a better name for it from my poor vocabulary.)

When it's `multiplePods`, PodAffinities can be satisfied by multiple Pods, whereas when it's `singlePod`, PodAffinities should be satisfied by a single Pod.

The disadvantage of it is that users cannot define like "want to use `satisfacatableBy: multiplePods` in `RequiredDuringSchedulingIgnoredDuringExecution`, but want to use `satisfacatableBy: singlePod` in  `PreferredDuringSchedulingIgnoredDuringExecution`".

### 2. a new field to have `[]PodAffinityTerm` that can be satisfied by multiple Pods.

Introduce two new `[]PodAffinityTerm` fields in [`PodAffinity`](https://pkg.go.dev/k8s.io/api/core/v1#PodAffinity) and those Affinities can be satisfied by multiple Pods. (One is for a required constraint, and the other is for a preferred constraint.)

Hmm, those fields' name... I haven't come up with good names to describe its behavior yet. 😓

**Comment by alculquicondor** [comment]
Exact names can be decided later, but working through the options in a KEP sounds like a good idea.

Do we have sufficient evidence that this feature is needed? In other words, are there real-world use cases where the feature was needed?

**Comment by Huang-Wei** [comment]
I'd suggest putting it on our plate only when there is a couple of viable practical use cases.

**Comment by sanposhiho** [comment]
Quoting from the [request](https://github.com/kubernetes/kubernetes/issues/68701#issuecomment-1257397866):
> Suppose we got an incoming pod with 2 pod affinity requirements. One is region and the other is zone. E.g the first one is for affinity to some light-weight RPC service it depends on, while the second for affinity to some heavy-weight storage access.

It makes sense. To be more general, when people want to make sure dependent services exist in the same domain and the pod has more than one kind of dependent services (like the above example), then they cannot use today’s PodAffinity to achieve that.

**Comment by alculquicondor** [comment]
I know it makes sense, but there doesn't seem to be enough pressure to get this done. This issue has been opened since 2018, and there aren't more people asking for it. Maybe you can collect some feedback sending an email to the mailing list?

**Comment by sanposhiho** [comment]
👍 In sig-scheduling mailing list? or is there another suitable one?

**Comment by Huang-Wei** [comment]
yes, and discuss.k8s.io

**Comment by fentas** [comment]
Curious shouldn't this be possible by reading the [design proposal](https://github.com/kubernetes/design-proposals-archive/blob/main/scheduling/podaffinity.md#algorithm)?
Also, it is nowhere written that you can't target multiple pods (or I over-/missread).

But it seems not, so my use case is related to:
- https://github.com/cilium/cilium/issues/19788
- https://github.com/Azure/AKS/issues/2934

We're running cilium in chaining with aks CSI. But since mid-last year removal of taints is only possible via Azure API breaking the cilium node readiness.
Right now, we're unable to run AKS in BYOCNI mode, unless it leaves its preview phase with Cilium.

Following [this comment from Microsoft](https://github.com/Azure/AKS/issues/2934#issuecomment-1122837301) we followed the advice and created a mutation webhook (instead of calling the API) adding a podAffinity so that pods only get scheduled on nodes where cilium agent is already running. 
But this breaks now other podAffinities.

**Comment by kerthcet** [comment]
Sorry, a bit confusion here, doesn't this works as expected? Unless the cilium pod launches, this new created pod will stuck in pending?

**Comment by alculquicondor** [comment]
@kerthcet the problem is that kube-scheduler looks for a single pod that satisfies all affinities.

Well, even if we add the feature that you request, it would only be available in 1.28 at the earliest (potentially 1.29, as the feature has to be disabled by default first). So I would suggest you open an issue against Azure support.

Other than that, you are welcome to work on this feature.
